File: 350_IntersectionOfTwoArraysII.cpp
//49.54%
class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        vector<int> res;
		unordered_map<int, int> table;
		for(auto num : nums1){
			if(table.find(num) == table.end())
				table[num] = 0;
			else
				table[num]++;
		}
		for(auto num : nums2){
			if(table.find(num) != table.end()){
				res.push_back(num);
				if(table[num])
					table[num]--;
				else
					table.erase(num);
			}	
		}
		return res;
    }
};

class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int, int> dict;
        vector<int> res;
        for(int i = 0; i < (int)nums1.size(); i++) dict[nums1[i]]++;
        for(int i = 0; i < (int)nums2.size(); i++)
            if(--dict[nums2[i]] >= 0) res.push_back(nums2[i]);
        return res;
    }
};


//Use two pointers and sort two vectors first
class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        sort(nums1.begin(), nums1.end());
        sort(nums2.begin(), nums2.end());
        int n1 = (int)nums1.size(), n2 = (int)nums2.size();
        int i1 = 0, i2 = 0;
        vector<int> res;
        while(i1 < n1 && i2 < n2){
            if(nums1[i1] == nums2[i2]) {
                res.push_back(nums1[i1]);
                i1++;
                i2++;
            }
            else if(nums1[i1] > nums2[i2]){
                i2++;
            }
            else{
                i1++;
            }
        }
        return res;
    }
};

File: 350_IntersectionOfTwoArraysII.py
# 77.2%
class Solution(object):
    def intersect(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[int]
        """
        import collections
        c = collections.Counter(nums1)
        res = list()
        for num in nums2:
            if num in c and c[num] > 0:
                c[num] -= 1
                res.append(num)
        return res

		
# A solution from LeetCode and pretty clear		
from collections import Counter

class Solution(object):
    def intersect(self, nums1, nums2):
        c1, c2 = Counter(nums1), Counter(nums2)
        return sum([[num] * min(c1[num], c2[num]) for num in c1 & c2], [])